### ISSUE TITLE
Segmentation fault issue possibly due to Keras module in TF2.0

### ISSUE THREAD

**Issue by nicolemitchell** [materialsvirtuallab/megnet#110]
I built off of your MEGNet implementation to create a node classification model.  I've been getting occasional Segmentation Faults while running hyperparameter tuning on my model. It does not happen consistently and the stack traces are not the same, though they seem to be triggered from the call to `fit_generator` in `train_from_graphs`.

Here's are stack traces from a few of these seg faults:
![Screen Shot 2020-03-29 at 6 36 33 PM](https://user-images.githubusercontent.com/15185137/77864746-051b8200-71f0-11ea-81e1-6a8d856c53fa.png)

![Screen Shot 2020-03-29 at 6 22 10 PM](https://user-images.githubusercontent.com/15185137/77864877-dce05300-71f0-11ea-9340-764c2e3fe9fa.png)

![Screen Shot 2020-03-29 at 6 20 47 PM](https://user-images.githubusercontent.com/15185137/77864904-00a39900-71f1-11ea-9cd2-372863d94c12.png)

I think this could be an issue with Keras being imported as a standalone module, while there is a tf.keras module in Tensorflow 2.0. When examining the stack trace, I see a mix of both modules: `python3.6/site-packages/tensorflow_core/python/keras/backend.py` and  `python3.6/site-packages/keras/engine/training.py`.

Is there a reason that you are still using the standalone Keras? I'm attempting to switch to tf.keras to eliminate any versioning issues, and so that I can build the MEGNet model dynamically and use Tensorflow 2.0's eager computation to more easily debug. However, this has turned out to not be trivial.

Furthermore, the following fatal error has also come up:

![Screen Shot 2020-03-29 at 6 26 38 PM](https://user-images.githubusercontent.com/15185137/77864489-62aecf00-71ee-11ea-8ef2-fec5c2043567.png)

I'm unsure if this is related to the Keras/tf.keras issue or a separate bug. It seems odd that the segment ids would not be increasing, as the bond indices are ordered when the graph is built.

I would appreciate any insight!
-Nicole

**Comment by chc273** [comment]
Hi Nicole @nicolemitchell 

Thanks for your interest in our work. 

The megnet codebase was written originally using tensorflow 1.x, and was recently upgraded to be compatible with tensorflow 2.0.0 in its main functionalitites.  We have not updated all APIs to tensorflow 2.0 yet (e.g., tf.keras and eager mode), since I have noticed issues with tensorflow 2.x in my other works using tensorflow 2.0 during the past months. Notable issues include performance degradation in running in eager mode and also occasional out-of-memory problems.

However, current megnet should not have such problems. We are also using megnet with tensorflow 2.0.0 with no problems. The seg fault problems you described above seem to be related to memory? 

For the seg id problems, I am not sure where the problem is. It seems that you are not using molecular data? I have not tested data format other than molecules and crystals, maybe there are some caveats in working with other data. I'd be interested in sorting this out if more details are provided. Do you have a minimal reproducible codebase?

Thanks

Chi

**Comment by chc273** [comment]
If you are using only part of megnet together with tensorflow2.x APIs and eager mode instead of using keras, it could also result in the problems above.

**Comment by nicolemitchell** [comment]
Hi Chi @chc273 

Thanks for your prompt reply!

> If you are using only part of megnet together with tensorflow2.x APIs and eager mode instead of using keras, it could also result in the problems above.

I am not using any Tensorflow 2.0 APIs or eager mode. However, I've noticed that the stack trace has calls to Tensorflow's eager mode. Is this expected? Or could it be from how Tensorflow and Keras are configured on my machine? 

> It seems that you are not using molecular data? I have not tested data format other than molecules and crystals, maybe there are some caveats in working with other data. I'd be interested in sorting this out if more details are provided. Do you have a minimal reproducible codebase?

I am using MEGNet on molecules. As input, I provide the Pybel molecular structures. The target I provide is a vector for each molecule of length num_atoms with binary labels. My model simply modifies the existing MEGNet model to feed only the node embeddings to the final dense layers and return a per-atom classification. Instead of the Set2Set and Concatenation layers following the MEGNet blocks, I create final_vec as follows:

`final_vec = Lambda(lambda x: x[0] + 0*x[1][:,0,:])([x1_, x3_])`

**Comment by chc273** [comment]
@nicolemitchell Thanks.  This is definitely an interesting topic to explore, but without a minimal example, I cannot really guess what's going on. 

Since it involves modifications of the codes and building new models, it seems to be more of a research project instead of an issue of megnet. Currently, we are also exploring possibilities to use megnet in molecular applications.

If you have a working example or a repo, I may be interested to see if I can build off it and include new models to the package.

**Comment by chc273** [comment]
@nicolemitchell  I
